app/services/physical_infra_topology_service.rb

```diff
@@ -78,16 +100,18 @@ def entity_status(entity)
     end
   end
 
-  # Decide whether or not to add an entity to the graph stack
-  def add_to_graph?(entity, links_index, links, parent_id)
-    return true if entity.kind_of?(Tag) || entity.kind_of?(PhysicalSwitch)
-
-    idx = links_index[entity_id(entity)]
-    # If a node has already been processed, change its parent rather than pushing to the stack
-    if idx
-      links[idx][:source] = parent_id
-      return false
+  # We don't allow a node to have more than one parent.
+  # The Physical Infra Topology must be a Tree without double links to the same node.
+  def filter_links(links)
+    has_parent = {}
+    filtered_links = []
+    links.reverse_each do |link|
+      target = link[:target]
+      unless has_parent[target]
+        has_parent[target] = true
+        filtered_links << link
+      end
     end
-    true
+    filtered_links
```

Comment:
```
has_parent = {}
links.reverse.each_with_object([]) do |link|
  target = link[:target]
  return if has_parent[target]
  has_parent[target] = true
  filtered_links << link
end
```

One less variable and one less level of nesting. Maybe?

Reply:
I wonder if we should have a spec for this method. Especially if it already had a bug in it. But I'd not delay merging this PR with that. You added a bunch of tests already.

Reply:
The tests I created are considering this `filter_links` in order to work properly.

In the test case, the entities would have more links than the result after the filtering. In the following graph, the edges in red are being removed by the `filter_links`:
![topology](https://user-images.githubusercontent.com/14334253/43459653-8befc1f4-94a4-11e8-8954-6e0568666e21.png)